digitalearthau/coherence.py

```diff
@@ -40,63 +47,121 @@ def main(expressions, archive_locationless, debug, check_locationless, check_anc
+                    locationless_count += 1
+
                     if archive_locationless:
                         dc.index.datasets.archive([dataset.id])
                         archive_count += 1
+                        _LOG.info("locationless_dataset_id.archived", dataset_id=str(dataset.id))
+                    else:
+                        _LOG.info("locationless_dataset_id", dataset_id=str(dataset.id))
 
             # If an ancestor is archived, it may have been replaced. This one may need
             # to be reprocessed too.
-            if check_ancestors:
-                _check_ancestors(check_siblings, dc, dataset)
+            if check_ancestors or archive_siblings or check_siblings:
+                archive_count += _check_ancestors(check_ancestors, check_siblings, archive_siblings, dc, dataset)
+
+        _LOG.info("coherence.finish",
+                  datasets_count=count,
+                  locationless_count=locationless_count,
+                  siblings_count=_siblings_count,
+                  archived_count=archive_count)
+
 
-        _LOG.info("coherence.finish", count=count, archive_count=archive_count)
+def _archive_dataset_ids(dc, ds_id, sibling_ids):
+    # sibling/s indexed_time dictionary
+    sb_itime_dict = dict()
 
+    # Count of dataset id's archived
+    archive_count = 0
 
-def _check_ancestors(check_siblings: bool, dc: Datacube, dataset: Dataset):
+    # dataset indexed time
+    ds_itime = dc.index.datasets.get(ds_id).indexed_time
+    for key, s_id in enumerate(sibling_ids):
+        sb_itime_dict[s_id] = dc.index.datasets.get(s_id).indexed_time
+
+    # new dataset id to retain
+    new_id = ds_id
+    for id, itime in sb_itime_dict.items():
+        # Check if dataset indexed time is less than or equal to sibling dataset indexed time
+        if ds_itime <= itime:
+            ds_id_archive = new_id
+            ds_archive_time = ds_itime
+
+            # Latest sibling dataset id to retain
+            new_id = id
+            ds_itime = itime
+        else:
+            ds_id_archive = id
+            ds_archive_time = itime
+
+        dc.index.datasets.archive([ds_id_archive])
+        _LOG.info("\tarchived_dataset_id", id=str(ds_id_archive), indexed_time=str(ds_archive_time))
+        archive_count += 1
+
+    _LOG.info("\trecent_dataset_id", id=str(new_id), indexed_time=str(ds_itime))
+    return archive_count
+
+
+def _check_ancestors(check_ancestors: bool,
+                     check_siblings: bool,
+                     archive_siblings: bool,
+                     dc: Datacube,
+                     dataset: Dataset):
+    global _siblings_count
+    ancestors_archive_count = 0
     dataset = dc.index.datasets.get(dataset.id, include_sources=True)
     if dataset.sources:
         for classifier, source_dataset in dataset.sources.items():
-            if source_dataset.is_archived:
+            if check_ancestors and source_dataset.is_archived:
```

Comment:
Would it make sense to not pass `check_ancestors` in here, and do this check regardless?

If the ancestor is _archived_ we probably don't care about duplicate siblings and would want to just archive them all.

Reply:
Removed check_ancestors check in the if condition